src/v/model/async_adl_serde.cc

```diff
@@ -32,4 +34,44 @@ async_adl<model::record_batch_reader>::from(iobuf_parser& in) {
       });
 }
 
+ss::future<>
+async_adl<model::record_batch>::to(iobuf& out, model::record_batch&& batch) {
+    batch_header hdr{
+      .bhdr = batch.header(),
+      .is_compressed = static_cast<int8_t>(batch.compressed() ? 1 : 0)};
+    reflection::serialize(out, hdr);
+    if (batch.compressed()) {
+        // This path isn't really async: we don't have a list to
+        // iterate over, just a single buffer to read.
+        reflection::serialize(out, std::move(batch).release_data());
+        return ss::now();
+    } else {
+        return batch.for_each_record_async(
```

Comment:
common use cases like this will be a user-after-free

```
auto batch = make_batch();
return async_adl<record_batch>{}.to(std::move(batch));
```

looks like either this needs to be a co_await and the parameter be a value, or you need to move the batch into a do_with?

Reply:
This is a do_with now